### PR TITLE
FEA Add `tabular_learner` factory function

### LINKED ISSUE #871: Handle numerical missing values in TableVectorizer
### Problem Description

Missing values are not handle by default

A reproducer:

```python
from skrub import TableVectorizer
from sklearn.pipeline import make_pipeline
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import fetch_openml


X_df, y = fetch_openml("titanic", version=1, as_frame=True, return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(X_df, y)

model = make_pipeline(TableVectorizer(), RandomForestClassifier())
model.fit(X_train, y_train).score(X_test, y_test)

```

Gives

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[1], line 12
      9 X_train, X_test, y_train, y_test = train_test_split(X_df, y)
     11 model = make_pipeline(TableVectorizer(), RandomForestClassifier())
---> 12 model.fit(X_train, y_train).score(X_test, y_test)

File ~/.local/miniconda/lib/python3.10/site-packages/sklearn/pipeline.py:405, in Pipeline.fit(self, X, y, **fit_params)
    403     if self._final_estimator != "passthrough":
    404         fit_params_last_step = fit_params_steps[self.steps[-1][0]]
--> 405         self._final_estimator.fit(Xt, y, **fit_params_last_step)
    407 return self

File ~/.local/miniconda/lib/python3.10/site-packages/sklearn/ensemble/_forest.py:345, in BaseForest.fit(self, X, y, sample_weight)
    343 if issparse(y):
    344     raise ValueError("sparse multilabel-indicator for y is not supported.")
--> 345 X, y = self._validate_data(
    346     X, y, multi_output=True, accept_sparse="csc", dtype=DTYPE
    347 )
    348 if sample_weight is not None:
    349     sample_weight = _check_sample_weight(sample_weight, X)

File ~/.local/miniconda/lib/python3.10/site-packages/sklearn/base.py:584, in BaseEstimator._validate_data(self, X, y, reset, validate_separately, **check_params)
    582         y = check_array(y, input_name="y", **check_y_params)
    583     else:
--> 584         X, y = check_X_y(X, y, **check_params)
    585     out = X, y
    587 if not no_val_X and check_params.get("ensure_2d", True):

File ~/.local/miniconda/lib/python3.10/site-packages/sklearn/utils/validation.py:1106, in check_X_y(X, y, accept_sparse, accept_large_sparse, dtype, order, copy, force_all_finite, ensure_2d, allow_nd, multi_output, ensure_min_samples, ensure_min_features, y_numeric, estimator)
   1101         estimator_name = _check_estimator_name(estimator)
   1102     raise ValueError(
   1103         f"{estimator_name} requires y to be passed, but the target y is None"
   1104     )
-> 1106 X = check_array(
   1107     X,
   1108     accept_sparse=accept_sparse,
   1109     accept_large_sparse=accept_large_sparse,
   1110     dtype=dtype,
   1111     order=order,
   1112     copy=copy,
   1113     force_all_finite=force_all_finite,
   1114     ensure_2d=ensure_2d,
   1115     allow_nd=allow_nd,
   1116     ensure_min_samples=ensure_min_samples,
   1117     ensure_min_features=ensure_min_features,
   1118     estimator=estimator,
   1119     input_name="X",
   1120 )
   1122 y = _check_y(y, multi_output=multi_output, y_numeric=y_numeric, estimator=estimator)
   1124 check_consistent_length(X, y)

File ~/.local/miniconda/lib/python3.10/site-packages/sklearn/utils/validation.py:921, in check_array(array, accept_sparse, accept_large_sparse, dtype, order, copy, force_all_finite, ensure_2d, allow_nd, ensure_min_samples, ensure_min_features, estimator, input_name)
    915         raise ValueError(
    916             "Found array with dim %d. %s expected <= 2."
    917             % (array.ndim, estimator_name)
    918         )
    920     if force_all_finite:
--> 921         _assert_all_finite(
    922             array,
    923             input_name=input_name,
    924             estimator_name=estimator_name,
    925             allow_nan=force_all_finite == "allow-nan",
    926         )
    928 if ensure_min_samples > 0:
    929     n_samples = _num_samples(array)

File ~/.local/miniconda/lib/python3.10/site-packages/sklearn/utils/validation.py:161, in _assert_all_finite(X, allow_nan, msg_dtype, estimator_name, input_name)
    144 if estimator_name and input_name == "X" and has_nan_error:
    145     # Improve the error message on how to handle missing values in
    146     # scikit-learn.
    147     msg_err += (
    148         f"\n{estimator_name} does not accept missing values"
    149         " encoded as NaN natively. For supervised learning, you might want"
   (...)
    159         "#estimators-that-handle-nan-values"
    160     )
--> 161 raise ValueError(msg_err)

ValueError: Input X contains NaN.
RandomForestClassifier does not accept missing values encoded as NaN natively. For supervised learning, you might want to consider sklearn.ensemble.HistGradientBoostingClassifier and Regressor which accept missing values encoded as NaNs natively. Alternatively, it is possible to preprocess the data, for instance by using an imputer transformer in a pipeline or drop samples with missing values. See https://scikit-learn.org/stable/modules/impute.html You can find a list of all estimators that handle NaN values at the following page: https://scikit-learn.org/stable/modules/impute.html#estimators-that-handle-nan-values
```

### Feature Description

if missing value in numeric, use `SimpleImputer`

### Alternative Solutions

_No response_

### Additional Context

_No response_

### LINKED ISSUE #924: adding `make_learner` to create a default pipeline for a given predictor
### Problem Description

As discussed IRL, it might be useful to have a function to help configure a simple pipeline that chains a TableVectorizer with a predictor chosen by the user.
The function would choose appropriate TableVectorizer parameters depending on the predictor -- for example MinHashEncoder is suitable for HistGradientBoostingClassifier but not for LogisticRegression

rough outline of the behavior:

<details>

```python
from sklearn.base import BaseEstimator
from sklearn import ensemble
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import OrdinalEncoder

from skrub import TableVectorizer, MinHashEncoder, ToCategorical


def make_learner(predictor, n_jobs=None):
    vectorizer = TableVectorizer(n_jobs=n_jobs)
    match predictor:
        case "classifier":
            return make_learner(
                ensemble.HistGradientBoostingClassifier(
                    categorical_features="from_dtype"
                ),
                n_jobs=n_jobs,
            )
        case "regressor":
            return make_learner(
                ensemble.HistGradientBoostingRegressor(
                    categorical_features="from_dtype"
                ),
                n_jobs=n_jobs,
            )
        case ensemble.HistGradientBoostingClassifier(
            categorical_features="from_dtype"
        ) | ensemble.HistGradientBoostingRegressor(categorical_features="from_dtype"):
            vectorizer.set_params(
                low_cardinality_transformer=ToCategorical(),
                high_cardinality_transformer=MinHashEncoder(),
            )
        case (
            ensemble.HistGradientBoostingClassifier()
            | ensemble.HistGradientBoostingRegressor()
            | ensemble.RandomForestClassifier()
            | ensemble.RandomForestRegressor()
        ):
            vectorizer.set_params(
                low_cardinality_transformer=OrdinalEncoder(),
                high_cardinality_transformer=MinHashEncoder(),
            )
        case BaseEstimator():
            pass
        case cls if issubclass(cls, BaseEstimator):
            raise TypeError("pass an estimator instance not the class")
        case _:
            raise TypeError("pass a scikit-learn estimator")

    # TODO: should we add an imputer when the model does not have built-in
    # support for missing values?
    return make_pipeline(vectorizer, predictor)
```

</details>

Usage

```python
>>> make_learner("regressor")
Pipeline(steps=[('tablevectorizer',
                 TableVectorizer(high_cardinality_transformer=MinHashEncoder(),
                                 low_cardinality_transformer=ToCategorical())),
                ('histgradientboostingregressor',
                 HistGradientBoostingRegressor(categorical_features='from_dtype'))])

>>> from sklearn.ensemble import HistGradientBoostingClassifier
>>> make_learner(HistGradientBoostingClassifier(), n_jobs=4)
Pipeline(steps=[('tablevectorizer',
                 TableVectorizer(high_cardinality_transformer=MinHashEncoder(),
                                 low_cardinality_transformer=OrdinalEncoder(),
                                 n_jobs=4)),
                ('histgradientboostingclassifier',
                 HistGradientBoostingClassifier())])

>>> from sklearn.linear_model import Ridge
>>> make_learner(Ridge())
Pipeline(steps=[('tablevectorizer', TableVectorizer()), ('ridge', Ridge())])
```

### Feature Description

...

### Alternative Solutions

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by jeromedockes**
closes #924 

also maybe enough to close #871 and #866?

**TODO:** 
- [ ] decide on name

### CODE REVIEW

**Comment by GaelVaroquaux** [comment]
Name suggestion: "make_skrub_learner" ?

We could have "make_skrub_pipeline", but I like to put the emphasis on the fact that this gives a supervised learner

**Comment by jeromedockes** [comment]
I don't have a strong opinion on the name. I think the advantage of having "pipeline" or something similar in it is it makes a clearer distinction between the learner provided by the user and the learner returned to them

**Comment by jeromedockes** [comment]
let's vote to pick a name in #938 , don't hesitate to add options!

File: examples/01_encodings.py [hunk 1]
```diff
@@ -241,6 +263,9 @@
 # Therefore it can be a faster and very effective alternative, when the supervised
 # learner is built on top of decision trees, which is the case for the |HGBR|.
 
+# The resulting pipeline is identical to the one produced by default by
```
Comment (TheooJ) [inline examples/01_encodings.py:266]: This is interpreted as a comment in a code block, if it's on purpose, I'd add it just after the `pipeline=make_pipeline[...]`

File: skrub/_make_tabular_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    This pipeline can handle complex, tabular data:
+
+    >>> import pandas as pd
+    >>> X = pd.DataFrame({'a': ['2020-01-02', '2021-04-01'], 'b': [None, 11], 'c': ['a', 'b']})
+    >>> y = [True, False]
+    >>> X
+                a     b  c
+    0  2020-01-02   NaN  a
+    1  2021-04-01  11.0  b
+    >>> model = make_tabular_pipeline('regressor').fit(X, y)
```
Comment (TheooJ) [inline skrub/_make_tabular_pipeline.py:69]: I'd probably do
```suggestion
    >>> model = make_tabular_pipeline('classifier').fit(X, y)
```
with a little more date (the current one predicts `[False, False]`)

File: skrub/_make_tabular_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    >>> model = make_tabular_pipeline(Ridge(solver='lsqr'))
+    >>> model.fit(X, y)
+    Pipeline(steps=[('tablevectorizer', TableVectorizer()),
+                    ('simpleimputer', SimpleImputer()),
+                    ('ridge', Ridge(solver='lsqr'))])
+
+    Here, different choices were made for the :obj:`TableVectorizer` -- in
+    particular the categorical column ``"c"`` is one-hot encoded, because the
+    :obj:`~sklearn.linear_model.Ridge` regressor lacks the
+    :obj:`~sklearn.ensemble.HistGradientBoostingRegressor`'s bult-in support
```
Comment (TheooJ) [inline skrub/_make_tabular_pipeline.py:121]:
```suggestion
    :obj:`~sklearn.ensemble.HistGradientBoostingRegressor`'s built-in support
```

File: skrub/_make_tabular_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    for categorical variables.
+
+    >>> model.named_steps['tablevectorizer'].transform(X)
+       a_year  a_month  a_day  a_total_seconds     b  c_b
+    0  2020.0      1.0    2.0     1.577923e+09   NaN  0.0
+    1  2021.0      4.0    1.0     1.617235e+09  11.0  1.0
+
+    Moreover, as :obj:`~sklearn.linear_model.Ridge` does not handle missing
+    values, a step was added to perform mean imputation. Therefore the data
+    seen by the final predictor actually looks like this (note scikit-learn
```
Comment (TheooJ) [inline skrub/_make_tabular_pipeline.py:131]:
```suggestion
    seen by the final predictor actually looks like this (note: scikit-learn
```

File: skrub/_make_tabular_pipeline.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+                n_jobs=n_jobs,
+            )
+        case "regressor":
+            return make_tabular_pipeline(
+                ensemble.HistGradientBoostingRegressor(**cat_feat_kwargs),
+                n_jobs=n_jobs,
+            )
+        case str():
+            raise ValueError(
+                "If ``predictor`` is a string it should be 'regressor' or 'classifier'"
```
Comment (TheooJ) [inline skrub/_make_tabular_pipeline.py:159]:
```suggestion
                "If ``predictor`` is a string it should be 'regressor' or 'classifier'."
```

**Review by TheooJ — APPROVED** [review]
LGTM ! We just need a name :)

File: skrub/_make_tabular_pipeline.py (hunk 2 above)
Comment (TheooJ) [reply skrub/_make_tabular_pipeline.py:69]: Could be 

```
>>> import pandas as pd
>>> from sklearn.linear_model import LogisticRegression

>>> X = pd.DataFrame(
    {'last_visit': ['2020-01-02', '2021-04-01', '2024-12-05', '2023-08-10'],
     'insulin_prescriptions': [None, 13, 0, 17],
     'fasting_glucose': [35, 140, 44, 137]}
)
>>> y = [False, True, False, True]

>>> model = make_tabular_pipeline(LogisticRegression())
>>> model.fit(X, y)
>>> model.predict(X)
array([False,  True, False,  True])

```

File: examples/01_encodings.py (hunk 1 above)
Comment (jeromedockes) [reply examples/01_encodings.py:266]: nope not on purpose :)

**Comment by GaelVaroquaux** [comment]
> +    >>> model = make_tabular_pipeline('regressor').fit(X, y)

When I see this, I feel that "skrub_learner" is a good name :)

**Comment by GaelVaroquaux** [comment]
> LGTM ! We just need a name :)

I'm also happy merging before the final name and doing the renaming. One benefit is that it will help people get a better idea of what this object does. Anyhow, we should settle on a name soon.

**Comment by TheooJ** [comment]
@GaelVaroquaux do we merge `table_learner` and make another PR if the name changes ?

**Comment by GaelVaroquaux** [comment]
> @GaelVaroquaux do we merge table_learner and make another PR if the name changes ?

I'd say: go for it. Use "tabular_learner" for now, as it's the one that gets the most positive vibes.

File: examples/01_encodings.py [hunk 6]
```diff
@@ -72,7 +72,29 @@
...
+#
+# - a |TableVectorizer| to preprocess the dataframe and vectorize the features
+# - a supervised learner (by default a |HGBR|)
+#
+# In the rest of this example, we focus on the first step and explore the
+# capabilities of skrub's |TableVectorizer|.
+
+# %%
+# Encoding tabular data
+# ---------------------
```
Comment (GaelVaroquaux) [inline examples/01_encodings.py:97]:
```suggestion
# More details on encoding tabular data
# -------------------------------------
```

File: skrub/_tabular_learner.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+
+def tabular_learner(estimator, n_jobs=None):
+    """Get a simple machine-learning pipeline for tabular data.
+
+    Scikit-learn estimators such as ``LogisticRegression()`` or
+    ``RandomForestClassifier()`` expect their input to be numeric arrays. They
+    do not accept heterogeneous dataframes containing complex data such as
+    datetimes or strings. Moreover, they do not always accept the input to
+    contain missing values. Therefore, some preprocessing must be applied to
+    dataframes before they are passed to an estimator.
```
Comment (GaelVaroquaux) [inline skrub/_tabular_learner.py:33]: The description on top of the docstring is too long. It should be moved to a "Notes" section, leaving only a quick read before the parameters section.

**Review by GaelVaroquaux — COMMENTED** [review]
This looks great. A few suggestions.

Can you also add it in  see also section in the docstring of the TableVectorizer, as well as on the encoding narrative doc (first section).

Thanks!!

**Comment by jeromedockes** [comment]
@GaelVaroquaux  IIRC some experimentation you're working on reveals that adding a missing value indicator when imputing usually helps. Should we set `add_indicator=True` on the `SimpleImputer` here?

**Comment by glemaitre** [comment]
> Should we set add_indicator=True on the SimpleImputer here?

Oh yes ;)

**Comment by GaelVaroquaux** [comment]
> Should we set add_indicator=True on the SimpleImputer here?

Yes, IMHO

File: skrub/_tabular_learner.py [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+                ensemble.HistGradientBoostingClassifier(**cat_feat_kwargs),
+                n_jobs=n_jobs,
+            )
+        if estimator == "regressor":
+            return tabular_learner(
+                ensemble.HistGradientBoostingRegressor(**cat_feat_kwargs),
+                n_jobs=n_jobs,
+            )
+        raise ValueError(
+            "If ``predictor`` is a string it should be 'regressor' or 'classifier'."
```
Comment (TheooJ) [inline skrub/_tabular_learner.py:197]:
```suggestion
            "If ``estimator`` is a string it should be 'regressor' or 'classifier'."
```

Comment (jeromedockes) [reply skrub/_tabular_learner.py:197]: thanks :)

**Comment by TheooJ** [comment]
Thanks @jeromedockes !

**Comment by glemaitre** [comment]
Oups, I was not fast enough for giving my review. I'll open a PR with little changes in the doc :)

**Comment by tomMoral** [comment]
great feature! thx :)

**Pull request merged**